aten/src/ATen/mps/MPSAllocator.h

```diff
@@ -0,0 +1,242 @@
+//  Copyright © 2022 Apple Inc.
+
+#include <ATen/Tensor.h>
+#include <ATen/ATen.h>
+#include <ATen/Utils.h>
+#include <torch/library.h>
+
+#include <ATen/mps/MPSDevice.h>
+#include <cstdio>
+#include <mutex>
+#include <unordered_map>
+#include <set>
+#include <utility>
+#include <mach/vm_page_size.h>
+
+#ifdef __OBJC__
+#include <Foundation/Foundation.h>
+#include <Metal/Metal.h>
+#include <Metal/MTLHeap.h>
+#include <MetalPerformanceShaders/MetalPerformanceShaders.h>
+#endif
```

Comment:
A high level description about the strategy behind this allocator would be much appreciated

Reply:
That's correct @ezyang. We have seen significant speedups by using a Caching allocator because MTLBuffer also has similar properties. We have kept the design same as CUDA caching allocator but the implementation is based on the MTLHeap implementation which does automatic placement for us.

We are relying on the Small and Large buffer heap concept from CUDA to leverage some of the tuning you guys have already done for PyTorch workloads. Mem allocator design can get very specific to the framework and workloads in question so we thought of using existing design would be a good start.

Reply:
Also added a high level description there.